src/item/enchantment/EnchantingTableOptionRegistry.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\item\enchantment;
+
+use pocketmine\item\enchantment\ItemEnchantmentTagRegistry as TagRegistry;
+use pocketmine\item\enchantment\ItemEnchantmentTags as Tags;
+use pocketmine\item\enchantment\VanillaEnchantments as Enchantments;
+use pocketmine\item\Item;
+use pocketmine\utils\SingletonTrait;
+use function array_filter;
+use function spl_object_id;
+
+/**
+ * Registry of enchantments that can be obtained using an enchanting table.
+ * Initially, it includes all enchantments from {@link VanillaEnchantments}, except "treasure" enchantments,
+ * such as Mending, Curse of Vanishing, Swift Sneak.
+ */
+final class EnchantingTableOptionRegistry{
+	use SingletonTrait;
+
+	/** @var Enchantment[] */
+	private array $enchantments = [];
+
+	/** @var string[][] */
+	private array $itemTags = [];
+
+	private function __construct(){
+		$this->register(Enchantments::PROTECTION(), [Tags::ARMOR]);
+		$this->register(Enchantments::FIRE_PROTECTION(), [Tags::ARMOR]);
+		$this->register(Enchantments::FEATHER_FALLING(), [Tags::BOOTS]);
+		$this->register(Enchantments::BLAST_PROTECTION(), [Tags::ARMOR]);
+		$this->register(Enchantments::PROJECTILE_PROTECTION(), [Tags::ARMOR]);
+		$this->register(Enchantments::THORNS(), [Tags::CHESTPLATE]);
+		$this->register(Enchantments::RESPIRATION(), [Tags::HELMET]);
+		$this->register(Enchantments::SHARPNESS(), [Tags::SWORD, Tags::AXE]);
+		$this->register(Enchantments::KNOCKBACK(), [Tags::SWORD]);
+		$this->register(Enchantments::FIRE_ASPECT(), [Tags::SWORD]);
+		$this->register(Enchantments::EFFICIENCY(), [Tags::DIG_TOOLS]);
+		$this->register(Enchantments::FORTUNE(), [Tags::DIG_TOOLS]);
+		$this->register(Enchantments::SILK_TOUCH(), [Tags::DIG_TOOLS]);
+		$this->register(Enchantments::UNBREAKING(), [Tags::ARMOR, Tags::DIG_TOOLS, Tags::SWORD, Tags::TRIDENT, Tags::BOW, Tags::CROSSBOW, Tags::FISHING_ROD]);
```

Comment:
`BREAKABLE`?

Reply:
No, Unbreaking in an enchanting table does not apply to all breakable items. BREAKABLE will be needed as a "secondary" Mending enchantment tag in the future

Reply:
which ones are omitted?

Reply:
Shears, Flint and Steel, Shield, Carrot on Stick, Elytra, Brush. They should all be in "secondary" tags for Unbreaking in the future

Reply:
Seems like BREAKABLE is useless then? Why aren't the secondary flags mentioned here?

Reply:
Yes, so far BREAKABLE is not used anywhere. I can remove it. What do you mean by not mentioning the secondary flags?

Reply:
> They should all be in "secondary" tags for Unbreaking in the future

Reply:
If you mean that EnchantingTableOptionRegistry doesn't specify secondary tags, it's simple: secondary tags are tags for items that can be enchanted not in an enchanting table, but in the anvil (or can be found on enchanted items in treasure chests). 
So for anvil, you'll have to then create an AnvilEnchantmentRegistry class where the secondary tags for enchantments are registered. And when checking the possibility of enchanting with an enchanted book in Anvil, both tags from AnvilEnchantmentRegistry and tags from EnchaningTableOptionRegistry will have to be checked. This is a consequence of the tags not being in Enchantment.

Reply:
Ugh, it seems like Mojang has a terrible name for everything in the enchanting system. So "primary flags" is actually "enchanting table enchantment flags" and "secondary flags" is actually "anvil enchantment flags".

Reply:
By the way, when implementing, for example, fishing or treasure chests, it may be difficult to get all available enchantments if the EnchantingTableOptionRegistry has enchantments only for the enchanting table, and the AnvilEnchantmentRegistry has only enchantments with "secondary" flags for anvil. You have to take from both registries and combine them, leaving only unique values in the array, and then check the tags from both registries... Are you sure it's ok to leave the system like this? Or maybe it would be better to create a general enchantment registry with primary and secondary tags at once, which would contain logic for both enchantment table and anvil and stuff?

Reply:
I'm having doubts about it myself. This is uncharted territory, so there's a good chance there's a better way to do this.

Reply:
I'll make an EnchantmentRegistry and you see how it is. I think it will still be better than creating separate registries

Reply:
Yeah, I'm not stoked about having two registries for this. Perhaps `AvailableEnchantmentRegistry`? Seems generic enough to cover all bases